### PR TITLE
[frontend] Skip access check for remote projects

### LINKED ISSUE #3587: EventMailer: Requests from remote projects
When submitting packages from remote projects to local projects `EventMailer` tries to generate a diff but that fails in a [access check](../blob/master/src/api/app/models/bs_request_action.rb#L922) because the it uses

```
Package.find_by_project_and_name(source_project, source_package)
```

which returns nil for remote projects. It the wanders off into the `if sp.nil?` and tries `User.-current.can_modify_project?` which fails miserably because there is no user logged in during delayed jobs.

### PR DESCRIPTION
**Pull request by hennevogel**
If a project is a remote project it is up to the instance that delivers the source to decide access rights.

Fixes #3587

### CODE REVIEW

File: src/api/app/models/bs_request_action.rb [hunk 1]
```diff
@@ -920,6 +920,8 @@ def expand_targets(ignore_build_state)
   end
 
   def source_access_check!
+    return if Project.is_remote_project?(source_project, true)
```
Comment (adrianschroeter) [inline src/api/app/models/bs_request_action.rb]: hm, did you benchmark this? IIRC it can be relative expensive, so we did this only when needed in the past

Comment (adrianschroeter) [inline src/api/app/models/bs_request_action.rb]: You can't hide the remote instance for certain users anymore by disabling access of the project which defines the remote instance when doing this.

To complete my former statement: It is up to the remote instance to deliver a package source or not. But this may still happen based on the credentials which are used to access in the remote instance. And they would be part of the local project defining the remote instance.

Eg.

   <remoteurl>https://adrian:secret@api.opensuse.org/public</remoteurl>

Comment (hennevogel) [inline src/api/app/models/bs_request_action.rb]: I have asked you exactly that question in the issue and you told me that it's up to the remote instance if it serves this or not....

The `remoteurl` you have pasted is in no way behaving differently in our `PublicController` then without credentials.

**Comment by adrianschroeter** [comment]
On Montag, 14. August 2017, 14:09:00 CEST wrote Henne Vogelsang:
> hennevogel commented on this pull request.
> 
> 
> 
> > @@ -920,6 +920,8 @@ def expand_targets(ignore_build_state)
>    end
>  
>    def source_access_check!
> +    return if Project.is_remote_project?(source_project, true)
> 
> I have asked you exactly that question in the issue and you told me that it's up to the remote instance if it serves this or not....

the small difference is that the delivery per package is decided by the remote instance, but the local instance
need to decide to make it available as a whole.

> The `remoteurl` you have pasted is in no way behaving differently in our `PublicController` then without credentials. 

right, it only works without the /public part.


-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by hennevogel** [comment]
> the small difference is that the delivery per package is decided by the remote instance, but the local instance need to decide to make it available as a whole.

So the local instance needs to decide this okay, then we are getting into many other questions as it's not always clear for which user this is in delayed jobs, especially not when sending mails...

**Comment by adrianschroeter** [comment]
well, the most simple solution here is just to use the nobody user to do the checking IMHO.

**Comment by hennevogel** [comment]
That is no solution that is a smear because it will just return false in doubt :-)

**Comment by hennevogel** [comment]
@adrianschroeter we talked about this again and we really would like to avoid introducing users to DJ and fixing all the problems resulting from it.

Would it be enough to disallow branching/diff generation from the local project (openSUSE.org) that defines the remote and let the remote instance decide for sources on it (openSUSE.org:openSUSE:Factory). This would hide your local project and it's remoteurl with credentials, but would allow branching.

**Comment by hennevogel** [comment]
@adrianschroeter ping

**Comment by adrianschroeter** [comment]
then do it to avoid loosing even more time with this, but document the regression in the release notes, since it can have a security impact to updaters. We may even should spit out a warning for them, if they have access control and remoteurl defined in a project.

(still the simple definition of setting the nobody user would avoid all these problems and it would be in sync what we do in the other jobs (but usually as Admin there) You could even ||= the nobody user in the source_access_check! method and avoid all this trouble).

**Comment by hennevogel** [comment]
@adrianschroeter I fear you don't get whats going on. Assigning `nobody` will just make `BsRequestAction.source_access_check!` crash further down. Let me explain 

```
  def source_access_check!
    sp = Package.find_by_project_and_name(source_project, source_package)
```
   `sp` will be nil for packages in remote projects.
```
    if sp.nil?
      # either not there or read permission problem
      if Package.exists_on_backend?(source_package, source_project)
```

The package might exist on the backend...

```
        # user is not allowed to read the source, but when he can write
        # the target, the request creator (who must have permissions to read source)
        # wanted the target owner to review it
        tprj = Project.find_by_name(target_project)
        if tprj.nil? || !User.current.can_modify_project?(tprj)
```
But `nobody` will never be able to modify `tprj`
```
          # produce an error for the source
          Package.get_by_project_and_name(source_project, source_package)
```
So we run this which just returns `nil` with no effect...
```
        end
        return
```
And if nobody could write to it we would return `nil` anyway...
```
      end
      if Project.exists_by_name(source_project)
```
The project always exists if it's below a `Project` defining a remote...
```
        # it is a remote project
        return
```
And we just return nil  to `RequestSourceDiff.gather_source_packages` which means we allow generating the diff.
```
      end
      # produce the same exception for webui
      Package.get_by_project_and_name(source_project, source_package)
```
If it doesn't exist this will just produce `nil` with no effect.
```
    end
    if sp.class == String
```
`nil` is never a `String`
```
      # a remote package
      return
    end
    sp.check_source_access!
```
And then we just call `check_source_access!` on `nil` which will crash
```
end
```

So either we allow generating the diff for everybody or we crash. This method just never was able to handle remote projects, it always crashed in some way. The difference to before is that we now see the exceptions. You will not get around deciding what to do in this case, then we can implement that decision...

**Comment by hennevogel** [comment]
I've split the commit into two, as 242e394921ee67365c665cf4d3ae06e44bbde70d is a functional change...

**Comment by hennevogel** [comment]
@adrianschroeter what now?

**Comment by hennevogel** [comment]
@adrianschroeter over 9k exceptions later.... Can you please give this some priority and review this again, either giving it a go or not....

**Comment by adrianschroeter** [comment]
if it crashes on sp.check_source_access! check this should maybe moved up into  the    if sp.class == String      case.

However, I extended the testcase test/functional/interconnect_test.rb  test_submit_requests_from_remote   locally with request diff calls and it worked. Isn't that also using the same code? What is different?

**Comment by hennevogel** [comment]
> What is different?

You are logged in...

**Comment by hennevogel** [comment]
dum di dum. 10.000...

**Comment by adrianschroeter** [comment]
and you would be logged in when the anonymous user is set, so I don't see how this would not solve the problem.

**Comment by bgeuken** [comment]
When the email is supposed to be sent to tux and Mr. Gecko, the permission check needs to be done for those two (since they will be the ones that see the source diff).
Checking the anonymous user won't help, and probably would fail also.

**Comment by hennevogel** [comment]
As explained above, depending on if the package exists on the backend it will either allow it for everybody (none of the rules apply) or crash...

**Comment by hennevogel** [comment]
@adrianschroeter can you please give this some priority?

**Comment by adrianschroeter** [comment]
For one I keep objecting against just skipping checks in doubt (not only remote, but also (source) access check when no user is set), this is not what I call "written with security in mind".

However, it is IMHO not very good that we have much special code at all for request permission checking, so I refactored this one in https://github.com/openSUSE/open-build-service/pull/3654. It is not solving your problem here, but we have IMHO a better base for discussion.

For solving the problem that source diffs might be sent to people who don't have access to it we have only two options IMHO (and this is independent of the remote case):
* Check source access for each person we send to the notification
* Simply assume nobody user to stay on the save side.

But please have a look in https://github.com/openSUSE/open-build-service/pull/3654 first please. It is IMHO a better foundation for further discussions since this code here is gone ...

**Comment by hennevogel** [comment]
First of all: Nothing is different with the changes in #3654. `source_access_check!` would call `BsRequestAction.is_target_maintainer?` without a `User.current` which will crash. 

> For one I keep objecting against just skipping checks in doubt (not only remote, but also (source) access check when no user is set), this is not what I call "written with security in mind".

We are not doing that. We explicitly allow diff generation for remote projects ( like *openSUSE.org:openSUSE:Factory*) we do not allow it to `Project`s with  a `remoteurl`! (like *openSUSE.org*) instead of doing the same implicit by assigning `nobody`. 

> For solving the problem that source diffs might be sent to people who don't have access to it we have only two options IMHO (and this is independent of the remote case):
> * Check source access for each person we send to the notification

Which we would need to implement. Shall we? That is what we are asking you for days now....

> Simply assume nobody user to stay on the save side.

Dude I'm telling you one **last** time: This is the same as allowing it for everybody if the package exists on the backend or crashing if the package does not exist on the backend. Can you please get over it?

**Comment by evanrolfe** [comment]
I'm finding it a bit hard to follow alll the posts in this conversation but from my point of view it seems that the original code was like this:
```ruby
if Project.exists_by_name(source_project)
  # it is a remote project
  return
end
```

So clearly `source_access_check!` was meant to return if the project is remote no? That is what @adrianschroeter wrote originally in this [commit](https://github.com/opensuse/open-build-service/commit/0f77200763e7044c0d0f6ac559bcf694be2e314a).

However that code fails for remote projects (as described in issue https://github.com/openSUSE/open-build-service/issues/3587) because that if block is never reached if the project is remote. This is causing ~11k errors on IBS so what we are doing in this PR is fixing that code to work as it was originally intended.

However @adrianschroeter it seems that you are now saying that `source_access_check!` should behave differently now that henne has submitted this PR, which is fair enough, but since we are just fixing a bug in the code which is currently causing lots and lots of errors to be raised, why don't we merge this PR as it is and then if you think this method should behave differently you can create a trello card so the change can go through the feature process (grooming,planning etc.) since clearly this discussion is PR going nowhere.. Just an idea anyway.

**Comment by adrianschroeter** [comment]
On Mittwoch, 23. August 2017, 13:14:11 CEST wrote Evan Rolfe:
> I'm finding it a bit hard to follow alll the posts in this conversation but from my point of view it seems that the original code was like this:
> ```ruby
> if Project.exists_by_name(source_project)
>   # it is a remote project
>   return
> end
> ```
> 
> So clearly `source_access_check!` was meant to return if the project is remote no? That is what @adrianschroeter wrote originally in this [commit](https://github.com/opensuse/open-build-service/commit/0f77200763e7044c0d0f6ac559bcf694be2e314a).

You don't need to check the remote package object, because this is the job of the remote instance.
But the local definition of the remote instance should be checked for read access.

> However that code fails for remote projects (as described in issue https://github.com/openSUSE/open-build-service/issues/3587) because that if block is never reached if the project is remote. This is causing ~11k errors on IBS so what we are doing in this PR is fixing that code to work as it was originally intended.
>
> However @adrianschroeter it seems that you are now saying that `source_access_check!` should behave differently now that henne has submitted this PR, which is fair enough, but since we are just fixing a bug in the code which is currently causing lots and lots of errors to be raised, why don't we merge this PR as it is and then if you think this method should behave differently you can create a trello card so the change can go through the feature process (grooming,planning etc.) since clearly this discussion is PR going nowhere.. Just an idea anyway.

Yes, if you just do the remote check (and ideally also check for read access check of the local object defining remote) you can do so.

However, I like to drop this entire code entirely since it is mostly a duplicate of the generic access check.

The only exception here is that it is okay to read the object defined in action_source if you
are maintainer of the target. Only this should be done here and otherwise run the generic code.

But again, for proper checking you need to run as matching user here. For checking the local
project in remote case or for checking the access check the local (action_source and action_target)
objects in other cases.

Skipping the checks for remote projects is just hiding the symptom (crash in delayed), but it
is no cure for the real problem.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by evanrolfe** [comment]
> Skipping the checks for remote projects is just hiding the symptom (crash in delayed), but it
is no cure for the real problem.

Ok so since this is a live bug occuring very frequently on IBS can we merge this PR and then come up with a proper solution in the next sprint?

**Comment by bgeuken** [comment]
> adrianschroeter commented
>
> You don't need to check the remote package object, because this is the job of the remote instance. But the local definition of the remote instance should be checked for read access.

So we would have to run the source access check for the project pointing to the remote?

But even if we do this, the source check is flawed when it's triggered by delayed jobs. It's not checking the actual user that will get the mail.

And we won't be able to fix this as a hotfix.

> evanrolfe commented
>>  > adrianschroeter commented
>>    Skipping the checks for remote projects is just hiding the symptom (crash in delayed), but it
>>    is no cure for the real problem.
>
> Ok so since this is a live bug occuring very frequently on IBS can we merge this PR and then come up with a proper solution in the next sprint?

That sounds like a good plan.

One alternative that would come to my mind is to disable notifications for remote projects and packages. That should silence the errors as well, right?
I guess you won't like this solution either, but one death we gone have to die here;-)

**Comment by evanrolfe** [comment]
@adrianschroeter please reply.. Do you accept the proposed solution of merging this PR and then implementing a more "proper" solution via the sprint process?

**Comment by hennevogel** [comment]
\> 12.000....

**Comment by DavidKang** [comment]
@adrianschroeter, ping!

**Comment by hennevogel** [comment]
20.000 exceptions now...

**Comment by DavidKang** [comment]
@adrianschroeter, ping!

**Comment by bgeuken** [comment]
I created a PR according to what we agreed on in the discussion we had after our standup. You can find it here -> #3697
